Enable parse_url for SparkSQL

codyschierbeck opened the pull request:
Copied over from #6143

May need to re-structure code for PrestoSQL parse_url to use a limited version of this extended version. Need to do path-finding first

codyschierbeck commented:
@PHILO-HE @FelixYBW Can you take a look?

PHILO-HE commented on velox/functions/sparksql/tests/URLFunctionsTest.cpp:
```diff
@@ -0,0 +1,222 @@
+ * limitations under the License.
+ */
+#include <gmock/gmock.h>
+#include <optional>
```
Seems a redundant `include`.

codyschierbeck replied:
Resolved

PHILO-HE reviewed (commented):
@codyschierbeck, could you document this function in a `rst` file? Thanks!

PHILO-HE commented on velox/functions/sparksql/URLFunctions.h:
```diff
@@ -0,0 +1,311 @@
+  }
+};
+
+static const std::unordered_set<submatchEnum> requiresAuthority = {
```
-> requiredXXX?

codyschierbeck replied:
No longer present in implementation

codyschierbeck commented:
Thank you for the review @PHILO-HE, will take a look

codyschierbeck commented:
@mbasmanova This is ready for review. I believe it is failing currently due to an issue you are addressing with #9606 

A couple of design clarifications: 

1) For intermediate processing, we leverage std::string_views often. Below is an example of why.

```cpp
    // [Userinfo@]Host[:Port].
    else if (partToExtract == "AUTHORITY") {
      if (!hasAuthority) {
        return false;
      }
      std::string_view host = submatch<std::string_view>(authorityMatch, kHost);
      std::string_view first = host;
      std::string_view last = host;

      std::string_view username = submatch<std::string_view>(authorityMatch, 1);
      std::string_view port = submatch<std::string_view>(authorityMatch, 4);
      if (!username.empty()) {
        first = username;
      }
      if (!port.empty()) {
        last = port;
      }
      result.setNoCopy(StringView(first.data(), last.end() - first.begin()));
      return true;
    }
```
This function uses "reuse_strings_from_arg = 0" because any valid result will always be present within the passed arg. 
We also know that if a result is made up of multiple matches those matches will be adjacent. In the above case, if userinfo, host, and port all exist within the URL we can merge the StringViews using the start of the "first" string_view and the distance between that point and the end of the "last" string_view. Unfortunately, this capability conflicts with velox::StringView in that if a string is inlined within the StringView, it is not stored in adjacent data anymore.

This leads to an interesting question: How does `result.setNoCopy()` work when the passed StringView is inlined? 

2) We use a static const regex that only compiles once.

3) There is some overlap between Presto and Spark parse_url. For example, the function "isValidURI" checks for whitespace and proper usage of [Percent-encoded URI’s](https://www.rfc-editor.org/rfc/rfc3986#section-2.1). There may be room in a future PR to find the overlapping functionality between the two implementations and bring them up to the velox/functions libraries.

4) Finally, I previously had an implementation that leveraged a switch statement and enums for the "partToExtract", hoping to squeeze out some performance enhancements since the chained if statement is performing a lot of string comparisons. I leveraged a custom micro-benchmarking tool I have locally that collects CPU FlameGraphs of the two implementations and found the results to be negligible (and at times worse using the switch and enums). 

   > 4a) I generally assume switch statements outperform chained string comparisons, so I am not sure why this is the case. My initial guess is that some compile-time optimizations replace the comparisons, but it may be that the way I implemented converting the string to an enum wasn't optimal (I used a static dictionary lookup, which may be naive).


### Output matching

   On the side, I have been working on a tool that runs a test case in PySpark, converts the PySpark DataFrame into Velox-compatible encodings, and then compares the output of the two functions. In its current state, it is very rudimentary but it does a decent job at sanity checking. Below is the output of a test suite that copies the test cases introduced for this function and includes a few extra corner cases to check.

<img width="158" alt="image" src="https://github.com/facebookincubator/velox/assets/14911410/de1ef5d7-9108-41bc-bd5d-ecf764511512">

Please let me know if there are any changes you would like to see.

FelixYBW commented:
@PHILO-HE

mbasmanova commented on velox/docs/functions/spark/url.rst:
```diff
@@ -0,0 +1,92 @@
+    # Examples of url functions with Invalid URI's.
+
+    # Invalid URI due to whitespace
+    SELECT url_extract_path('foo '); -- NULL (1 row)
```
It doesn't look like there is url_extract_path in Spark. Would you please double check that this documentation makes sense for Spark?

codyschierbeck replied:
Resolved

mbasmanova commented on velox/docs/functions/spark/url.rst:
```diff
@@ -0,0 +1,92 @@
+
+    Extracts a part from a URL. The part to extract must be QUERY, otherwise function will return null:
+
+    :param string: The URL to extract the part from.
```
Would you generate the docs locally to make sure they render nicely?

codyschierbeck replied:
Docs are looking good locally.

<img width="421" alt="image" src="https://github.com/facebookincubator/velox/assets/14911410/b40d7968-3f22-4854-81d9-8c04d51e640b">

mbasmanova commented on velox/docs/functions/spark/url.rst:
```diff
@@ -0,0 +1,92 @@
+
+.. spark:function:: parse_url(string, partToExtract, key) -> varchar
+
+    Extracts a part from a URL. The part to extract must be QUERY, otherwise function will return null:
```
Why do we need partToExtract parameter if it can have only one value?

codyschierbeck replied:
Passing a partToExtract that isn't QUERY is possible, but it will return null. Below is output from sanity checker confirming:

<img width="208" alt="image" src="https://github.com/facebookincubator/velox/assets/14911410/eb74e0e1-005a-46ad-9138-187e9c7aa7b1">

mbasmanova commented on velox/functions/sparksql/URLFunctions.h:
```diff
@@ -0,0 +1,311 @@
+  }
+};
+
+} // namespace facebook::velox::functions::sparksql
```
+1

codyschierbeck replied:
On my local end this is solved and pushed, but checking the GitHub files, the empty line at the end isn't present. When I add a second newline to try and force GitHub to display it, `make format-fix` deletes the second newline.

mbasmanova reviewed (commented):
@codyschierbeck Some initial comments.

mbasmanova commented on velox/functions/sparksql/URLFunctions.h:
```diff
@@ -0,0 +1,310 @@
+#include "velox/functions/Macros.h"
+
+namespace facebook::velox::functions::sparksql {
+namespace {
```
no anonymous namespaces in header files ;use "detail" namespace

codyschierbeck replied:
Resolved

mbasmanova commented on velox/functions/sparksql/URLFunctions.h:
```diff
@@ -0,0 +1,310 @@
+namespace facebook::velox::functions::sparksql {
+namespace {
+
+/// Performs initial validation of the URI.
```
/// for public APIs, // for private

codyschierbeck replied:
Resolved. In this case, would that mean the comment above ParseUrlFunction should be /// ? As of now it is 

```cpp
// parse_url(url, partToExtract) → string
// parse_url(url, partToExtract, key) → string
//
// Extracts a part of a URL. The partToExtract argument can be one of
// 'PROTOCOL', 'HOST', 'PATH', 'REF', 'AUTHORITY', 'FILE', 'USERINFO',
// 'QUERY', 'FRAGMENT'.
//
// If the URL is invalid, return null.
template <typename T>
struct ParseUrlFunction {
```

mbasmanova commented on velox/functions/sparksql/URLFunctions.h:
```diff
@@ -0,0 +1,310 @@
+namespace {
+
+/// Performs initial validation of the URI.
+/// Checks if the URI contains ascii whitespaces or
```
The previous sentence seems redundant.

`Performs initial validation of the URI.`

codyschierbeck replied:
Resolved

mbasmanova commented on velox/functions/sparksql/URLFunctions.h:
```diff
@@ -0,0 +1,310 @@
+  buf[2] = '\0';
+  char* endptr;
+  for (; p < end; ++p) {
+    if (facebook::velox::functions::stringImpl::isAsciiWhiteSpace(*p)) {
```
can you drop 'facebook::velox::functions::'?

codyschierbeck replied:
Resolved

mbasmanova commented on velox/functions/sparksql/URLFunctions.h:
```diff
@@ -0,0 +1,310 @@
+  return true;
+}
+
+static const boost::regex kUriRegex(
```
why boost and not RE2?

codyschierbeck replied:
I chose to user boost to match Velox Proesto implementation to make room for consolidating implementations in a future PR. I will create a local implementation using RE2 and measure performance.

mbasmanova commented on velox/functions/sparksql/URLFunctions.h:
```diff
@@ -0,0 +1,310 @@
+}
+
+template <>
+FOLLY_ALWAYS_INLINE StringView submatch(const boost::cmatch& match, int idx) {
```
Would it make sense to for all 'submatch' functions to use std::string_view for return type?

codyschierbeck replied:
One reason this function is templated is found in the 3 parameter call function, which extracts the value of a key from a query. 

Below is that function, but it is modified with the detail namespace and a version of submatch that only returns std::string_view. We compare an `arg_type<Varchar>& key` (velox::StringView) to a `std::string_view k`

```cpp

    auto query = detail::submatch(match, kQuery);
    const boost::cregex_iterator begin(
        query.data(), query.data() + query.size(), kQueryParamRegex);
    boost::cregex_iterator end;

    for (auto it = begin; it != end; ++it) {
      if (it->length(2) != 0) { // key shouldn't be empty.
        auto k = detail::submatch((*it), 2);
        if (key.compare(k) == 0) {
          auto value = detail::submatch((*it), 3);
          if (value != "") {
            result.setNoCopy(StringView(value.data(), value.size()));
            return true;
          } else {
            return false;
          }
        }
      }
    }
```

In this version, we run into incompatibilities between std::string_view and StringView. First, key.compare(k) fails to compile because velox::StringView does not support comparisons against std::string_views. Second, result.setNoCopy does not support std::string_views, so we are forced to convert to StringView.

In this case it is thus easier to start with a StringView since we do not need a std::string_view.

A second reason, that will be included in the changes from this review, is when we know that we won't need to merge any StringViews (anytime simpleMatch is called), it is also better to avoid converting to a velox::StringView in the result.setNoCopy() call.

mbasmanova commented on velox/functions/sparksql/URLFunctions.h:
```diff
@@ -0,0 +1,310 @@
+  return boost::regex_match(
+      rawUrl.data(), rawUrl.data() + rawUrl.size(), match, kUriRegex);
+}
+// PARSE_URL(url, partToExtract) → string
```
add empty line; why all CAPS?

codyschierbeck replied:
Resolved

mbasmanova commented on velox/functions/sparksql/URLFunctions.h:
```diff
@@ -0,0 +1,310 @@
+// PARSE_URL(url, partToExtract, key) → string
+//
+// Extracts a part of a URL. The partToExtract argument can be one of
+// 'PROTOCOL', 'HOST', 'PATH', 'REF', 'AUTHORITY', 'FILE', 'USERINFO',
```
Do these have to be in all caps? Would be nice to clarify in the docs.

codyschierbeck replied:
Yes, they do have to be all capitalized. Below is output from my sanity checker confirming this:

<img width="191" alt="image" src="https://github.com/facebookincubator/velox/assets/14911410/4a92967e-f8ad-4c24-847b-bc229aad4fae">

Will update documentation to make this more clear

mbasmanova commented on velox/functions/sparksql/URLFunctions.h:
```diff
@@ -0,0 +1,310 @@
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+  static constexpr int kAuthPath = 3;
```
make these private

codyschierbeck replied:
Resolved

mbasmanova commented on velox/functions/sparksql/URLFunctions.h:
```diff
@@ -0,0 +1,310 @@
+  static constexpr int kPass = 2;
+  static constexpr int kPort = 4;
+
+  FOLLY_ALWAYS_INLINE bool call(
```
This method is very long. Let's not inline it and let's break it up into smaller pieces.

codyschierbeck replied:
Created private helper functions for path, file, userinfo, and authority

mbasmanova commented on velox/functions/sparksql/URLFunctions.h:
```diff
@@ -0,0 +1,310 @@
+  }
+
+ private:
+  FOLLY_ALWAYS_INLINE bool matchAuthorityAndPath(
```
remove inline

mbasmanova commented on velox/functions/sparksql/URLFunctions.h:
```diff
@@ -0,0 +1,310 @@
+    hasAuthority = true;
+    return true;
+  }
+  FOLLY_ALWAYS_INLINE bool simpleMatch(
```
add empty line

codyschierbeck replied:
Resolved

mbasmanova commented on velox/functions/sparksql/tests/URLFunctionsTest.cpp:
```diff
@@ -0,0 +1,222 @@
+}
+
+} // namespace
+} // namespace facebook::velox::functions::sparksql::test
```
+1

codyschierbeck replied:
Ditto above

mbasmanova reviewed (commented):
@codyschierbeck Please, address @PHILO-HE 's comments.

codyschierbeck commented:
Apologies for the reply spam.

@mbasmanova @PHILO-HE Please see comments addressing change requests.

### Summary

1. Removed inlining of both 'call' functions.
2. I created separate functions for all the partToExtract cases that come after the authority regex match (HOST, PATH, FILE, USERINFO, AUTHORITY).
3. Any cases that do not require std::string_view merging now use the StringView template of submatch.
4. Moved the submatch index static variables to the private section of the struct
5. Moved isValidURI, kUriRegex, submatch, and parse from a nameless namespace to the detail namespace
6. Removed full namespace path of isAsciiWhiteSpace in isValidURI.
7. Updated documentation for clarity and accuracy
8. Updated comments to match coding standards.
9. Addressed newline requests.
